src/transformers/models/whisper/modeling_whisper.py

```diff
@@ -289,66 +296,44 @@ def forward(
         bsz, tgt_len, _ = hidden_states.size()
 
         # get query proj
-        query_states = self.q_proj(hidden_states) * self.scaling
+        query_states = self.q_proj(hidden_states)
         # get key, value proj
         # `past_key_value[0].shape[2] == key_value_states.shape[1]`
         # is checking that the `sequence_length` of the `past_key_value` is the same as
         # the provided `key_value_states` to support prefix tuning
-        if (
-            is_cross_attention
-            and past_key_value is not None
-            and past_key_value[0].shape[2] == key_value_states.shape[1]
-        ):
-            # reuse k,v, cross_attentions
-            key_states = past_key_value[0]
-            value_states = past_key_value[1]
-        elif is_cross_attention:
```

Comment:
Here according to my understanding, key_states and value_states are once and for all computed based on encoder hidden states when we are doing cross-attention, so we can cache them ideally, but I think these are a little bit different from the existing caches because we don't need to update the cache in every generation step, we just have to do it once in the first step, so we could either try creating another Cache class and in that cases we need to pass in two caches(one for self attention and one for cross attention)? or just initiating and updating the cross attention cache within every layer like recurrent gemma?  What do you think is best @ArthurZucker

Reply:
Your understanding is correct: we compute the k/v states for the cross-attention **once** in the first forward pass, and then save them to cache. We would indeed want to cache the k/v states to avoid re-computing them at every step.

What do you think about the proposed design in this PR? https://github.com/huggingface/transformers/pull/28931#issue-2125652215 Similar to your solution, it uses two cache's: one for the self-attention, and one for the cross-attention. Note that the cache specifics are out-of-date, given the recent changes to the cache API, but the high-level design remains valid!

Reply:
Yeah, it would be great if two StaticCache would work, but the issue is in that case we can not tell whether we are in the first generation step or not because the shape of cache now is always fixed to hold the maximum generation size

Reply:
my current solution is use two StaticCache and add another boolen parameter in forward to mark whether we are doing the first generation step, this will cause a recapture in the second step but should work fine for the subsequent steps

Reply:
The current solution utilizes a new `OneShotCache`, and a tuple of two caches is expected for conditional generation